File: 147_Insertion_Sort_List/147.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def insertionSortList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        # fake start
        Dummy = ListNode(0)
        # head is the current node, move it till end(nullptr)
        while head is not None:
            # the processing node
            ptr = head
            # move head
            head = head.next
            # insertion place
            pos = Dummy
            # move insertion place when next node is not None and smaller than it
            while pos.next is not None and pos.next.val < ptr.val:
                pos = pos.next
            # pos->ptr->pos.next
            ptr.next = pos.next
            pos.next = ptr
        return Dummy.next
        

File: 147_Insertion_Sort_List/147.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* insertionSortList(ListNode* head) {
        // fake start
        ListNode dummy(0);
        // head is the current node, move it till end(nullptr)
        while(head != nullptr){
            // the processing node
            ListNode *ptr = head;
            // move head 
            head = head->next;
            // insertion place
            ListNode *pos = &dummy;
            // move insertion place when next node is not None and smaller than it
            while(pos->next != nullptr && pos->next->val < ptr->val)
                pos = pos->next;
            // pos->ptr->pos.next
            ptr->next = pos->next;
            pos->next = ptr;
        }
        return dummy.next;
    }
};